plugin/core/transports.py

```diff
@@ -87,24 +129,16 @@ def __del__(self) -> None:
     def _read_loop(self) -> None:
         try:
             while self._reader:
-                headers = http.client.parse_headers(self._reader)  # type: ignore
-                body = self._reader.read(int(headers.get("Content-Length")))
-                try:
-                    payload = _decode(body)
-
-                    def invoke(p: Dict[str, Any]) -> None:
-                        callback_object = self._callback_object()
-                        if callback_object:
-                            callback_object.on_payload(p)
-
-                    sublime.set_timeout_async(partial(invoke, payload))
-                except Exception as ex:
-                    exception_log("JSON decode error", ex)
+                payload = self._processor.read_data(self._reader)
+                if payload is None:
                     continue
-                finally:
-                    # We don't need these anymore
-                    del body
-                    del headers
+
```

Comment:
This is probably not very generic.

I've noticed that this depends on an exception being thrown to exit the loop when the process has exited. But that logic seems to be fairly specific to JSON RPC since with a processor that just reads lines as text there won't normally be any exception triggered.

Ideally there should probably be a more explicit way to shutdown the thread...